test/unit/tools/unified_spec_runner.test.ts

```diff
@@ -0,0 +1,345 @@
+import { BSON } from 'bson';
+import { AssertionError, expect } from 'chai';
+
+import { EntitiesMap } from '../../tools/unified-spec-runner/entities';
+import { compareLogs, resultCheck } from '../../tools/unified-spec-runner/match';
+import { ExpectedLogMessage } from '../../tools/unified-spec-runner/schema';
+
+describe('Unified Spec Runner', function () {
+  describe('Matching', function () {
+    describe('resultCheck', function () {
+      const entitiesMap: EntitiesMap = new EntitiesMap();
+      let ranResultCheck: boolean;
+      let actual: any;
+      let expected: any;
+
+      const runResultCheck = () => {
+        ranResultCheck = true;
+        resultCheck(actual, expected, entitiesMap, []);
+      };
+
+      beforeEach(function () {
+        ranResultCheck = false;
+      });
+
+      afterEach(function () {
+        expect(ranResultCheck, 'Test cannot pass unless resultCheck has been called').to.be.true;
+      });
+
+      context('$$matchAsDocument', function () {
+        beforeEach(function () {
+          expected = {
+            $$matchAsDocument: {
+              data: { $$exists: true },
+              a: { $$type: ['int'] }
+            }
+          };
+        });
+
+        context('when actual value is EJSON string', function () {
+          it('throws AssertionError when it finds extra keys', function () {
+            actual =
+              '{"data": {"$numberLong": "100"}, "a": {"$numberInt": "10"}, "b": {"$numberInt": "100"}}';
+            expect(runResultCheck).to.throw(AssertionError);
```

Comment:
Is there a reason we're relying on the `runResultCheck` helper?  Usually we just inline the function call in an arrow function:  `expect(() => resultCheck(actual, expected, entitiesMap, [])).to.throw(AssertionError)`.

Also, we need to specifically assert that `resultCheck` fails for the correct reasons.  Can we add a regular expression match as well? `expect(() => ....).to.throw(AssertionError, /regex here/)`

Reply:
The reason we're relying on the `runResultCheck` helper is just to make sure that no test can pass unless the function under test has been called. If you take a look at the top-level `beforeEach` and `afterEach` hooks, you can see where that's done.

Reply:
Does `expect(() => ...).not.to.throw(AssertionError)` not cover the passing cases?

Reply:
It would run, but my rationale behind putting that helper is just to ensure that if a new test is added that the only way it can pass is if it calls the function in question. I can go ahead and remove it since it was more of a development convenience than anything.

Reply:
My thought is that it's probably not necessary, but extra testing never hurts.  If we do want to enforce this with after hooks, I'd suggest using sinon.
```typescript
import * as sinon from 'sinon';

const resultCheckSpy = sinon.spy(resultCheck);

afterEach(() => {
  expect(resultCheckSpy.called).to.be.true;
});
```